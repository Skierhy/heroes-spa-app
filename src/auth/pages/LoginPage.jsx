import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../context/auth/AuthContext';

export const LoginPage = () => {
	const navigate = useNavigate();
	// dar acceso al useContext
	const { login } = useContext(AuthContext);
	const onLogin = () => {
		const lastPath = localStorage.getItem('lastPath') || '/';
		login('Skierhy');
		navigate(lastPath, { replace: true });
	};
	return (
		<>
			<div className=''>
				<div>
					<h1>Login</h1>
					<hr />
					<button className='btn btn-primary' onClick={onLogin}>
						Login
					</button>
				</div>
			</div>
		</>
	);
};
