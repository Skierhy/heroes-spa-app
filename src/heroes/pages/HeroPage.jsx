import { useMemo } from 'react';
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroesById } from '../helpers';

export const HeroPage = () => {
	// useParams() es un hook de react-router-dom que nos permite obtener los parámetros de la url
	const { id } = useParams();
	// useMemo() es un hook de react que nos permite memorizar un valor, en este caso el resultado de la función getHeroesById
	// si el id cambia, se vuelve a ejecutar la función
	const hero = useMemo(() => getHeroesById(id), [id]);
	// useNavigate() es un hook de react-router-dom que nos permite navegar a una ruta
	const navigate = useNavigate();
	const onNavigateBack = () => {
		navigate(`/${hero.publisher === 'DC Comics' ? 'dc' : 'marvel'}`);
		// otra forma de hacerlo
		// navigate(-1);
	};

	const styles = {
		color: '#ffffff',
	};

	// esta condición es para que si no existe el id del héroe, redirija a la pagina de heroes
	if (!hero) return <Navigate to='/' />;
	return (
		<>
			<div className='row mt-5 animate__animated animate__flipInX'>
				<div className='col-xl-4 col-sm-12'>
					<img
						src={`/assets/heroes/${id}.jpg`}
						alt={hero.superhero}
						className='img-thumbnail'
					/>
				</div>
				<div className='col-xl-8 col-sm-12 mt-3'>
					<h3>{hero.superhero}</h3>
					<ul className='list-group list-group-flush'>
						<li className='list-group-item'>
							<b>Alter ego: </b>
							{hero.alterEgo}
						</li>
						<li className='list-group-item'>
							<b>Publisher: </b>
							{hero.publisher}
						</li>
						<li className='list-group-item'>
							<b>First appearance: </b>
							{hero.firstAppearance}
						</li>
					</ul>
					<h5 className='mt-3' style={styles}>
						Characters
					</h5>
					<p style={styles}>{hero.characters}</p>
					<button
						className='btn btn-outline-info'
						onClick={onNavigateBack}
					>
						Return
					</button>
				</div>
			</div>
		</>
	);
};
