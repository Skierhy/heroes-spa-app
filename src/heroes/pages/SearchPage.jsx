import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../components';
import { getHeroesByName } from '../helpers';

const styles = {
	color: '#ffffff',
};

export const SearchPage = () => {
	// useNavigate es un hook que nos permite navegar a una ruta
	const navigate = useNavigate();
	// useLocation es un hook que nos permite obtener la ruta actual
	const location = useLocation();
	// siempre recibe un string
	const { q = '' } = queryString.parse(location.search);
	const heroes = getHeroesByName(q);
	const { searchText, onInputChange } = useForm({
		searchText: q,
	});
	const onSearchSubmit = (e) => {
		e.preventDefault();
		if (searchText.trim().length <= 1) return;
		// en esta linea se hace la navegaciÃ³n
		// a la ruta /search?query=batman
		navigate(`?q=${searchText}`);
	};
	return (
		<>
			<h1>Search</h1>
			<hr />
			<div className='row'>
				<div className='col-xl-5 col-sm-12'>
					<h4 style={styles}>Searching</h4>
					<hr />
					<form onSubmit={onSearchSubmit}>
						<input
							type='text'
							placeholder='Find your hero'
							className='form-control'
							name='searchText'
							autoComplete='off'
							value={searchText}
							onChange={onInputChange}
						/>
						<button
							type='submit'
							className='btn mt-3 btn-block btn-outline-primary'
						>
							Search...
						</button>
					</form>
				</div>
				<div className='col-xl-7 col-sm-12'>
					<h4 style={styles} className='mt-2'>
						Results
					</h4>
					<hr />
					{q === '' ? (
						<div className='alert alert-info'>Search a hero</div>
					) : (
						heroes.length === 0 && (
							<div className='alert alert-danger'>
								There is no a hero with <b>{q}</b>
							</div>
						)
					)}
					{heroes.map((hero) => (
						<HeroCard key={hero.id} {...hero} />
					))}
				</div>
			</div>
		</>
	);
};
