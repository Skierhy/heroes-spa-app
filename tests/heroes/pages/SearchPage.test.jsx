import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <SearchPage />', () => {
	beforeEach(() => jest.clearAllMocks());

	test('debe de mostrarse correctamente con valores por defecto', () => {
		const { container } = render(
			<MemoryRouter>
				<SearchPage />
			</MemoryRouter>
		);
		// se harÃ¡ un snapshot del componente
		expect(container).toMatchSnapshot();
	});

	test('debe de mostrar a Batman y el input con el valor del queryString', () => {
		// initialEntries={['/search?q=batman']
		// es para simular la ruta actual
		// en este caso es la ruta /search?q=batman
		render(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<SearchPage />
			</MemoryRouter>
		);
		// se espera que el input tenga el valor batman
		const input = screen.getByRole('textbox');
		expect(input.value).toBe('batman');

		// debe la imagen debe que tener el atributo para mostrar la imagen
		const img = screen.getByRole('img');
		expect(img.src).toContain('/images/heroes/dc-batman.jpg');

		// se espera que el h5 tenga el valor batman
		const h5 = screen.getByText('Batman');
		expect(h5).toBeTruthy();
	});

	test('debe de mostrar un error si no se encuentra el hero (batman123)', () => {
		render(
			<MemoryRouter initialEntries={['/search?q=batman123']}>
				<SearchPage />
			</MemoryRouter>
		);
		// el alert debe que contener el texto 'No hay un hero con batman123'
		const alert = screen.getByLabelText('alert-danger');
		expect(alert.innerHTML).toContain('batman123');
	});

	test('debe de llamar el navigate a la pantalla nueva', () => {
		const inputValue = 'superman';

		render(
			<MemoryRouter initialEntries={['/search']}>
				<SearchPage />
			</MemoryRouter>
		);

		const input = screen.getByRole('textbox');
		// change es para simular el cambio de valor en el input
		fireEvent.change(input, {
			target: { name: 'searchText', value: inputValue },
		});

		const form = screen.getByRole('form');
		fireEvent.submit(form);
		// se espera que el navigate sea llamado con el valor superman
		expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
	});
});
