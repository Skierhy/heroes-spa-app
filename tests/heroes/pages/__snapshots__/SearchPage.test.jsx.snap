// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-xl-5 col-sm-12"
    >
      <h4
        style="color: rgb(255, 255, 255);"
      >
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-xl-7 col-sm-12"
    >
      <h4
        class="mt-2"
        style="color: rgb(255, 255, 255);"
      >
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
